SourceX/controls/game_controls.cpp

```diff
@@ -48,57 +46,79 @@ bool GetGameAction(const SDL_Event &event, GameAction *action)
 	const ControllerButtonEvent ctrl_event = ToControllerButtonEvent(event);
 	if (!InGameMenu()) {
 		switch (ctrl_event.button) {
-		case ControllerButton::IGNORE:
-			return true;
 		case ControllerButton::AXIS_TRIGGERLEFT: // ZL (aka L2)
 			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::TOGGLE_CHARACTER_INFO);
+				if (IsControllerButtonPressed(ControllerButton::BUTTON_BACK))
+					*action = GameAction(GameActionType::TOGGLE_QUEST_LOG);
+				else
+					*action = GameAction(GameActionType::TOGGLE_CHARACTER_INFO);
 			return true;
 		case ControllerButton::AXIS_TRIGGERRIGHT: // ZR (aka R2)
 			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::TOGGLE_INVENTORY);
-			return true;
-		case ControllerButton::BUTTON_B: // Right button
-			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::PRIMARY_ACTION);
-			return true;
-		case ControllerButton::BUTTON_Y: // Top button
-			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::SECONDARY_ACTION);
-			return true;
-		case ControllerButton::BUTTON_X: // Left button
-			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::CAST_SPELL);
+				if (IsControllerButtonPressed(ControllerButton::BUTTON_BACK))
+					*action = GameAction(GameActionType::TOGGLE_SPELL_BOOK);
+				else
+					*action = GameAction(GameActionType::TOGGLE_INVENTORY);
 			return true;
-		case ControllerButton::BUTTON_A: // Bottom button
-			if (!ctrl_event.up)
-				*action = GameAction(GameActionType::TOGGLE_QUICK_SPELL_MENU);
-			return true;
-		case ControllerButton::BUTTON_LEFTSHOULDER:
-			if (!stextflag && !ctrl_event.up)
-				*action = GameAction(GameActionType::USE_HEALTH_POTION);
-			return true;
-		case ControllerButton::BUTTON_RIGHTSHOULDER:
-			if (!stextflag && !ctrl_event.up)
-				*action = GameAction(GameActionType::USE_MANA_POTION);
-			return true;
-		case ControllerButton::BUTTON_DPAD_UP:
-		case ControllerButton::BUTTON_DPAD_DOWN:
-		case ControllerButton::BUTTON_DPAD_LEFT:
-		case ControllerButton::BUTTON_DPAD_RIGHT:
-			// The rest is handled in charMovement() on every game_logic() call.
-			return true;
-		case ControllerButton::BUTTON_RIGHTSTICK:
-			*action = GameActionSendMouseClick { GameActionSendMouseClick::LEFT, ctrl_event.up };
-			return true;
-		default:
-			break;
+		}
+		if (!questlog && !sbookflag) {
+			switch (ctrl_event.button) {
+			case ControllerButton::IGNORE:
+				return true;
+			case ControllerButton::BUTTON_B: // Right button
+				if (!ctrl_event.up)
+					if (IsControllerButtonPressed(ControllerButton::BUTTON_BACK))
+						*action = GameActionSendKey { DVL_VK_F8, ctrl_event.up };

```

Comment:
Should we be sending the down events for these as well or is it not necessary?

Reply:
We ONLY need to send the down event, which I think it what we are doing, at least it is working :)

Reply:
Right, we're, OK!